sdk/search/Azure.Search.Documents/src/Batching/Publisher{T}.cs

```diff
@@ -0,0 +1,493 @@
+        /// Add documents to be published.  This should only be called outside
+        /// of the Publisher itself.
+        /// </summary>
+        /// <param name="documents">The documents to publish.</param>
+        /// <param name="async">Whether to call sync or async.</param>
+        /// <param name="cancellationToken">Cancellation token.</param>
+        /// <returns>Task for adding the documents.</returns>
+        public async Task AddDocumentsAsync(IEnumerable<T> documents, bool async, CancellationToken cancellationToken = default)
+        {
+            EnsureNotDisposed();
+            await SendMessageInternal(Message.Publish(documents), async, cancellationToken).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Flush any pending documents.  This should only be called outside of
+        /// the Publisher itself.
+        /// </summary>
+        /// <param name="async">Whether to call sync or async.</param>
+        /// <param name="cancellationToken">Cancellation token.</param>
+        /// <returns>Task for adding the documents.</returns>
+        public async Task FlushAsync(bool async, CancellationToken cancellationToken = default)
+        {
+            await SendMessageInternal(Message.Flush(), async, cancellationToken).ConfigureAwait(false);
+            await _flushBarrier.WaitAsync(async, cancellationToken).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Send a message from a producer to the channel.
+        /// </summary>
+        /// <param name="message">The message to send.</param>
+        /// <param name="async">Whether to execute sync or async.</param>
+        /// <param name="cancellationToken">Cancellation token.</param>
+        /// <returns>Task for the sending the message.</returns>
+        private async Task SendMessageInternal(
+            Message message,
+            bool async,
+            CancellationToken cancellationToken = default)
+        {
+            // Send the message
+            if (async)
+            {
+                await _channel.Writer.WriteAsync(message, cancellationToken).ConfigureAwait(false);
+            }
+            else
+            {
+                // It's an unbounded channel so spinning should be fine
+                while (!_channel.Writer.TryWrite(message)) { /* spin */ }
+            }
+        }
+        #endregion
+
+        #region Message Consumer
+        /// <summary>
+        /// Listen for and process messages.
+        /// </summary>
+        /// <returns>
+        /// A Task that will run for the lifetime of the publisher.
+        /// </returns>
+        private async Task ProcessMessagesAsync()
+        {
+            CancellationToken cancellationToken = PublisherCancellationToken;
+            while (await _channel.Reader.WaitToReadAsync(cancellationToken).ConfigureAwait(false))
+            {
+                Message message = await _channel.Reader.ReadAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
`while (_channel.Reader.TryRead)`?

Reply:
That's the sync overload that doesn't block, right?  Wouldn't that exit as soon as the channel is empty?